### PR TITLE
fix(core): Bail out of atom.write when atom update and current value are equal

### LINKED ISSUE #877: [core] dependency tracking fails with derived chains
Split from: https://github.com/pmndrs/jotai/issues/874#issuecomment-987338240

Repro with sync atoms
https://codesandbox.io/s/jotai-demo-forked-cucwg

Repro with async atoms
https://codesandbox.io/s/jotai-demo-forked-y58e3

### PR DESCRIPTION
**Pull request by Thisen**
Fixes #877.

### CODE REVIEW

File: src/core/store.ts [hunk 1]
```diff
@@ -426,6 +426,10 @@ export const createStore = (
           // NOTE technically possible but restricted as it may cause bugs
           throw new Error('atom not writable')
         }
+        const aState = readAtomState(a)
```
Comment (dai-shi) [inline src/core/store.ts:429]: Can this be `getAtomState`? readAtomState tries to do more work.

**Review by dai-shi — COMMENTED** [review]
Great work!!!
So, setAtomPromiseOrValue bails out for non-promise values, but we are invoking invalidateDependents anyway, which causes this behavior. This is partly my intention, but overlooked the reported behavior.

File: src/core/store.ts (hunk 1 above)
Comment (Thisen) [reply src/core/store.ts]: Done 👍🏼

**Review by dai-shi — APPROVED** [review]
Thanks again for this! This is new to me and motivates me to investigate some more, because I feel like having some other bugs. I will try.

**Pull request merged**